Error NETSDK1144: Optimizing assemblies for size failed

### Description

Last night the builds just started failing - we hadn’t made any obvious changes which would affect it. It is failing on the dotnet publish step.

`dotnet publish $(solutionName) -f net9.0-android `

It fails running dotnet publish on my PC as well, although the build works fine in VS. 

The error is below (same for Android and iOS except for the code):

C:\Users\VssAdministrator.nuget\packages\microsoft.net.illink.tasks\9.0.6\build\Microsoft.NET.ILLink.targets(96,5): error NETSDK1144: Optimizing assemblies for size failed. [D:\a\1\s\NetCore\Client\PDSX\PDS.App.PDSX.csproj::TargetFramework=net9.0-android]

I have discovered the problem.

The build is failing because of the 'where' line of code below. With that code in place the build fails. If I remove that line of code the build succeeds.

```
var modalStack = Application.Current?.Windows[0].Navigation.ModalStack;

if (modalStack.Any())
{
    await Application.Current.Dispatcher.DispatchAsync(async () =>
    {
        var navigation = Application.Current?.Windows[0].Navigation;

        while (navigation.ModalStack.Any(page => page is TPage)) // <- THIS IS THE PROBLEM
        {
            // Find the topmost TPage and pop it
            for (int i = navigation.ModalStack.Count - 1; i >= 0; i--)
            {
                if (navigation.ModalStack[i] is TPage)
                {
                    await navigation.PopModalAsync(false);
                    await Task.Delay(300);
                    break;
                }
            }
        }
    });
}
```

### Steps to Reproduce

Insert above code
Execute dotnet publish command

### Link to public reproduction project repository

N/A

### Version with bug

9.0.70 SR7

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

9.0.70 SR7

### Affected platforms

iOS, Android

### Affected platform versions

Android, ios

### Did you find any workaround?

No

### Relevant log output

```shell
Android:

C:\Users\VssAdministrator.nuget\packages\microsoft.net.illink.tasks\9.0.6\build\Microsoft.NET.ILLink.targets(143,5): error MSB6006: “dotnet.exe” exited with code -1073741571. [D:\a\1\s\NetCore\Client\PDSX\PDS.App.PDSX.csproj::TargetFramework=net9.0-android]
C:\Users\VssAdministrator.nuget\packages\microsoft.net.illink.tasks\9.0.6\build\Microsoft.NET.ILLink.targets(96,5): error NETSDK1144: Optimizing assemblies for size failed. [D:\a\1\s\NetCore\Client\PDSX\PDS.App.PDSX.csproj::TargetFramework=net9.0-android]

iOS:
/Users/runner/.nuget/packages/microsoft.net.illink.tasks/9.0.6/build/Microsoft.NET.ILLink.targets(143,5): error MSB6006: "dotnet" exited with code 134. [/Users/runner/work/1/s/NetCore/Client/PDSX/PDS.App.PDSX.csproj::TargetFramework=net9.0-ios]
/Users/runner/.nuget/packages/microsoft.net.illink.tasks/9.0.6/build/Microsoft.NET.ILLink.targets(96,5): error NETSDK1144: Optimizing assemblies for size failed. [/Users/runner/work/1/s/NetCore/Client/PDSX/PDS.App.PDSX.csproj::TargetFramework=net9.0-ios]
```

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs
@@ -175,8 +175,8 @@ namespace ILCompiler.Dataflow
                                         // Find calls to state machine constructors that occur outside the type
                                         if (referencedMethod.IsConstructor &&
                                             referencedMethod.OwningType is MetadataType generatedType &&
-                                            // Don't consider calls in the same type, like inside a static constructor
-                                            method.OwningType != generatedType &&
+                                            // Don't consider calls in the same/nested type, like inside a static constructor
+                                            !IsSameOrNestedType(method.OwningType, generatedType) &&
                                             CompilerGeneratedNames.IsLambdaDisplayClass(generatedType.Name))
                                         {
                                             Debug.Assert(generatedType.IsTypeDefinition);
@@ -216,8 +216,8 @@ namespace ILCompiler.Dataflow
                                         field = field.GetTypicalFieldDefinition();
 
                                         if (field.OwningType is MetadataType generatedType &&
-                                            // Don't consider field accesses in the same type, like inside a static constructor
-                                            method.OwningType != generatedType &&
+                                            // Don't consider field accesses in the same/nested type, like inside a static constructor
+                                            !IsSameOrNestedType(method.OwningType, generatedType) &&
                                             CompilerGeneratedNames.IsLambdaDisplayClass(generatedType.Name))
                                         {
                                             Debug.Assert(generatedType.IsTypeDefinition);
@@ -258,6 +258,22 @@ namespace ILCompiler.Dataflow
                         // Already warned above if multiple methods map to the same type
                         // Fill in null for argument providers now, the real providers will be filled in later
                         generatedTypeToTypeArgs[stateMachineType] = new TypeArgumentInfo(method, null);
+                    }
+
+                    static bool IsSameOrNestedType(TypeDesc type, TypeDesc potentialOuterType)
+                    {
+                        do
+                        {
+                            if (type == potentialOuterType)
+                                return true;
+
+                            if (type is not EcmaType ecmaType)
+                                return false;
+
+                            type = ecmaType.ContainingType;
+                        } while (type != null);
+
+                        return false;
                     }
                 }
 

--- a/src/tools/illink/src/linker/Linker.Dataflow/CompilerGeneratedState.cs
+++ b/src/tools/illink/src/linker/Linker.Dataflow/CompilerGeneratedState.cs
@@ -180,8 +180,8 @@ namespace Mono.Linker.Dataflow
                                 // Find calls to state machine constructors that occur outside the type
                                 if (referencedMethod.IsConstructor &&
                                     referencedMethod.DeclaringType is var generatedType &&
-                                    // Don't consider calls in the same type, like inside a static constructor
-                                    method.DeclaringType != generatedType &&
+                                    // Don't consider calls in the same/nested type, like inside a static constructor
+                                    !IsSameOrNestedType(method.DeclaringType, generatedType) &&
                                     CompilerGeneratedNames.IsLambdaDisplayClass(generatedType.Name))
                                 {
                                     // fill in null for now, attribute providers will be filled in later
@@ -219,8 +219,8 @@ namespace Mono.Linker.Dataflow
                                     continue;
 
                                 if (field.DeclaringType is var generatedType &&
-                                    // Don't consider field accesses in the same type, like inside a static constructor
-                                    method.DeclaringType != generatedType &&
+                                    // Don't consider field accesses in the same/nested type, like inside a static constructor
+                                    !IsSameOrNestedType(method.DeclaringType, generatedType) &&
                                     CompilerGeneratedNames.IsLambdaDisplayClass(generatedType.Name))
                                 {
                                     if (!generatedTypeToTypeArgs.TryAdd(generatedType, new TypeArgumentInfo(method, null)))
@@ -252,6 +252,19 @@ namespace Mono.Linker.Dataflow
                     // Already warned above if multiple methods map to the same type
                     // Fill in null for argument providers now, the real providers will be filled in later
                     generatedTypeToTypeArgs[stateMachineType] = new TypeArgumentInfo(method, null);
+                }
+
+                static bool IsSameOrNestedType(TypeDefinition type, TypeDefinition potentialOuterType)
+                {
+                    do
+                    {
+                        if (type == potentialOuterType)
+                            return true;
+
+                        type = type.DeclaringType;
+                    } while (type != null);
+
+                    return false;
                 }
             }
 

--- a/src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/CompilerGeneratedTypes.cs
+++ b/src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/CompilerGeneratedTypes.cs
@@ -48,6 +48,8 @@ namespace Mono.Linker.Tests.Cases.DataFlow
             CapturingLocalFunctionInsideIterator<int>();
             LambdaInsideAsync<int>();
             LocalFunctionInsideAsync<int>();
+            NestedAsyncLambda.Test<int>();
+            NestedAsyncLocalFunction.Test<int>();
             NestedStaticLambda.Test<int>();
         }
 
@@ -421,6 +423,50 @@ namespace Mono.Linker.Tests.Cases.DataFlow
             void LocalFunction() => typeof(T).GetMethods();
             await Task.Delay(0);
             LocalFunction();
+        }
+
+        class NestedAsyncLambda
+        {
+            public static async void Test<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] T>()
+            {
+                var outer = async () =>
+                {
+                    var inner =
+                    [ExpectedWarning("IL2090", "T", nameof(DynamicallyAccessedMemberTypes.PublicProperties))]
+                    () =>
+                    {
+                        _ = typeof(T).GetMethods();
+                        _ = typeof(T).GetProperties();
+                    };
+
+                    inner();
+                };
+
+                outer().Wait();
+            }
+        }
+
+        class NestedAsyncLocalFunction
+        {
+            public static void Test<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] T>()
+            {
+                Local1();
+
+#if INCLUDE_UNEXPECTED_LOWERING_WARNINGS
+                [UnexpectedWarning("IL2091", "T", nameof(DynamicallyAccessedMemberTypes.PublicMethods), Tool.Trimmer | Tool.NativeAot, "https://github.com/dotnet/roslyn/issues/79333", CompilerGeneratedCode = true)]
+#endif
+                static async Task Local1()
+                {
+                    Local2();
+
+                    [ExpectedWarning("IL2090", "T", nameof(DynamicallyAccessedMemberTypes.PublicProperties))]
+                    static void Local2()
+                    {
+                        _ = typeof(T).GetMethods();
+                        _ = typeof(T).GetProperties();
+                    };
+                };
+            }
         }
 
         class NestedStaticLambda

